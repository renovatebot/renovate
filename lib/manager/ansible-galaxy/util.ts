export const newBlockRegEx = /^\s*-\s*((\w+):\s*(.*))$/;
export const blockLineRegEx = /^\s*((\w+):\s*(.*))$/;
export const galaxyDepRegex = /[\w-]+\.[\w-]+/;
export const dependencyRegex = /^dependencies:/;
export const galaxyRegEx =
  /^\s+(?<lookupName>[\w.]+):\s*["'](?<version>.+)["']\s*/;
export const nameMatchRegex =
  /(?<source>((git\+)?(?:(git|ssh|https?):\/\/)?(.*@)?(?<hostname>[\w.-]+)(?:(:\d+)?\/|:))(?<depName>[\w./-]+)(?:\.git)?)(,(?<version>[\w.]*))?/;
