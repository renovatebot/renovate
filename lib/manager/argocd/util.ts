export const keyValueExtractionRegex =
  /^\s*(?<key>[^\s]+):\s+"?(?<value>[^"\s]+)"?\s*$/;
// looks for `apiVersion: argoproj.io/
export const fileTestRegex = /\s*apiVersion:\s*argoproj.io\/\s*/;
