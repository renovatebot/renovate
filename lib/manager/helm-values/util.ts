export type HelmDockerImageDependency = {
  registry?: string;
  repository: string;
  tag: string;
};

/**
 * This is a workaround helper to allow the usage of 'unknown' in
 * a type-guard function while checking that keys exist.
 *
 * @see https://github.com/microsoft/TypeScript/issues/21732
 * @see https://stackoverflow.com/a/58630274
 */
function hasKey<K extends string>(k: K, o: {}): o is { [_ in K]: {} } {
  return typeof o === 'object' && k in o;
}

/**
 * Type guard to determine whether a given partial Helm values.yaml object potentially
 * defines a Helm Docker dependency.
 *
 * There is no exact standard of how Docker dependencies are defined in Helm
 * values.yaml files (as of January 1st 2020), this function defines a
 * heuristic based on the most commonly used format in the stable Helm charts:
 *
 * image:
 *   repository: 'something'
 *   tag: v1.0.0
 */
export function matchesHelmValuesDockerHeuristic(
  parentKey: string,
  data: unknown
): data is HelmDockerImageDependency {
  return (
    parentKey === 'image' &&
    data &&
    typeof data === 'object' &&
    hasKey('repository', data) &&
    hasKey('tag', data)
  );
}
