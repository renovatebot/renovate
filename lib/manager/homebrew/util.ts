export function skip(
  idx: number,
  content: string,
  cond: (s: string) => boolean
): number {
  let i = idx;
  while (i < content.length) {
    if (!cond(content[i])) {
      return i;
    }
    i += 1;
  }
  return i;
}

export function isSpace(c: string): boolean {
  return /\s/.test(c);
}

export function removeComments(content: string): string {
  let newContent = removeLineComments(content);
  newContent = removeMultiLineComments(newContent);
  return newContent;
}

// Remove line comments starting with #
function removeLineComments(content: string): string {
  let newContent = '';
  let comment = false;
  for (let i = 0; i < content.length; i += 1) {
    const c = content[i];
    if (c === '#') {
      comment = true;
    }
    if (comment) {
      if (c === '\n') {
        comment = false;
      }
    }
    if (!comment) {
      newContent += c;
    }
  }
  return newContent;
}

// Remove multi-line comments enclosed between =begin and =end
function removeMultiLineComments(content: string): string {
  const beginRegExp = /(^|\n)=begin\s/;
  const endRegExp = /(^|\n)=end\s/;
  let newContent = content;
  let i = newContent.search(beginRegExp);
  let j = newContent.search(endRegExp);
  while (i !== -1 && j !== -1) {
    if (newContent[i] === '\n') {
      i += 1;
    }
    if (newContent[j] === '\n') {
      j += 1;
    }
    j += '=end'.length;
    newContent = newContent.substring(0, i) + newContent.substring(j);
    i = newContent.search(beginRegExp);
    j = newContent.search(endRegExp);
  }
  return newContent;
}
