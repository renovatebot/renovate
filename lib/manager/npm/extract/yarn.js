const yarnLockParser = require('@yarnpkg/lockfile');
const { logger } = require('../../../logger');

module.exports = {
  getYarnLock,
};

async function getYarnLock(filePath) {
  const yarnLockRaw = await platform.getFile(filePath);
  try {
    const yarnLockParsed = yarnLockParser.parse(yarnLockRaw);
    // istanbul ignore if
    if (yarnLockParsed.type !== 'success') {
      logger.info(
        { filePath, parseType: yarnLockParsed.type },
        'Error parsing yarn.lock - not success'
      );
      return {};
    }
    const lockFile = {};
    for (const [entry, val] of Object.entries(yarnLockParsed.object)) {
      logger.trace({ entry, version: val.version });
      lockFile[entry] = val.version;
      // istanbul ignore if
      if (val.integrity) {
        lockFile['@renovate_yarn_integrity'] = true;
      }
    }
    return lockFile;
  } catch (err) {
    logger.info({ filePath, err }, 'Warning: Exception parsing yarn.lock');
    return {};
  }
}
