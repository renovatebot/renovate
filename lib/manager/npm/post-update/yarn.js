const fs = require('fs-extra');
const upath = require('upath');
const { getInstalledPath } = require('get-installed-path');
const { exec } = require('child-process-promise');

module.exports = {
  generateLockFile,
};

async function generateLockFile(cwd, env, config = {}, upgrades = []) {
  const { binarySource } = config;
  logger.debug(`Spawning yarn install to create ${cwd}/yarn.lock`);
  let lockFile = null;
  let stdout;
  let stderr;
  let cmd;
  try {
    const startTime = process.hrtime();
    try {
      // See if renovate is installed locally
      const installedPath = upath.join(
        await getInstalledPath('yarn', {
          local: true,
        }),
        'bin/yarn.js'
      );
      cmd = `node ${installedPath}`;
      const yarnIntegrity =
        config.upgrades &&
        config.upgrades.some(upgrade => upgrade.yarnIntegrity);
      if (!yarnIntegrity) {
        try {
          const renovatePath = await getInstalledPath('renovate', {
            local: true,
          });
          logger.info('Using nested bundled yarn@1.9.4 for install');
          cmd = 'node ' + upath.join(renovatePath, 'bin/yarn-1.9.4.js');
        } catch (err) {
          logger.info('Using bundled yarn@1.9.4 for install');
          cmd = cmd.replace(
            'node_modules/yarn/bin/yarn.js',
            'bin/yarn-1.9.4.js'
          );
        }
      }
    } catch (localerr) {
      logger.debug('No locally installed yarn found');
      // Look inside globally installed renovate
      try {
        const renovateLocation = await getInstalledPath('renovate');
        const installedPath = upath.join(
          await getInstalledPath('yarn', {
            local: true,
            cwd: renovateLocation,
          }),
          'bin/yarn.js'
        );
        cmd = `node ${installedPath}`;
      } catch (nestederr) {
        logger.debug('Could not find globally nested yarn');
        // look for global yarn
        try {
          const installedPath = upath.join(
            await getInstalledPath('yarn'),
            'bin/yarn.js'
          );
          cmd = `node ${installedPath}`;
        } catch (globalerr) {
          logger.warn('Could not find globally installed yarn');
          cmd = 'yarn';
        }
      }
    }
    if (binarySource === 'global') {
      cmd = 'yarn';
    }
    logger.debug(`Using yarn: ${cmd}`);
    let cmdExtras = '';
    cmdExtras += ' --ignore-scripts';
    cmdExtras += ' --ignore-engines';
    cmdExtras += ' --ignore-platform';
    cmdExtras += process.env.YARN_MUTEX_FILE
      ? ` --mutex file:${process.env.YARN_MUTEX_FILE}`
      : ' --mutex network:31879';
    const installCmd = cmd + ' install' + cmdExtras;
    // TODO: Switch to native util.promisify once using only node 8
    ({ stdout, stderr } = await exec(installCmd, {
      cwd,
      shell: true,
      env,
    }));
    logger.debug(`yarn stdout:\n${stdout}`);
    logger.debug(`yarn stderr:\n${stderr}`);
    const lockUpdates = upgrades
      .filter(upgrade => upgrade.isLockfileUpdate)
      .map(upgrade => upgrade.depName);
    if (lockUpdates.length) {
      logger.info('Performing lockfileUpdate (yarn)');
      const updateCmd =
        cmd +
        ' upgrade' +
        lockUpdates.map(depName => ` ${depName}`).join('') +
        cmdExtras;
      const updateRes = await exec(updateCmd, {
        cwd,
        shell: true,
        env,
      });
      stdout += updateRes.stdout
        ? /* istanbul ignore next */ updateRes.stdout
        : '';
      stderr += updateRes.stderr
        ? /* istanbul ignore next */ updateRes.stderr
        : '';
    }
    if (
      config.postUpdateOptions &&
      config.postUpdateOptions.includes('yarnDedupeFewer')
    ) {
      logger.info('Performing yarn dedupe fewer');
      const dedupeCommand =
        'npx yarn-deduplicate@1.1.1 --strategy fewer && yarn';
      const dedupeRes = await exec(dedupeCommand, {
        cwd,
        shell: true,
        env,
      });
      stdout += dedupeRes.stdout
        ? /* istanbul ignore next */ dedupeRes.stdout
        : '';
      stderr += dedupeRes.stderr
        ? /* istanbul ignore next */ dedupeRes.stderr
        : '';
    }
    if (
      config.postUpdateOptions &&
      config.postUpdateOptions.includes('yarnDedupeHighest')
    ) {
      logger.info('Performing yarn dedupe highest');
      const dedupeCommand =
        'npx yarn-deduplicate@1.1.1 --strategy highest && yarn';
      const dedupeRes = await exec(dedupeCommand, {
        cwd,
        shell: true,
        env,
      });
      stdout += dedupeRes.stdout
        ? /* istanbul ignore next */ dedupeRes.stdout
        : '';
      stderr += dedupeRes.stderr
        ? /* istanbul ignore next */ dedupeRes.stderr
        : '';
    }
    const duration = process.hrtime(startTime);
    const seconds = Math.round(duration[0] + duration[1] / 1e9);
    lockFile = await fs.readFile(upath.join(cwd, 'yarn.lock'), 'utf8');
    logger.info(
      { seconds, type: 'yarn.lock', stdout, stderr },
      'Generated lockfile'
    );
  } catch (err) /* istanbul ignore next */ {
    logger.info(
      {
        cmd,
        err,
        stdout,
        stderr,
        type: 'yarn',
      },
      'lock file error'
    );
    if (err.stderr) {
      if (err.stderr.includes('ENOSPC: no space left on device')) {
        throw new Error('disk-space');
      }
      if (
        err.stderr.includes('The registry may be down.') ||
        err.stderr.includes('getaddrinfo ENOTFOUND registry.yarnpkg.com') ||
        err.stderr.includes('getaddrinfo ENOTFOUND registry.npmjs.org')
      ) {
        throw new Error('registry-failure');
      }
    }
    return { error: true, stderr: err.stderr };
  }
  return { lockFile };
}
