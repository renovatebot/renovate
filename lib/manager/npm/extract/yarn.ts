import { parse } from '@yarnpkg/lockfile';
import { logger } from '../../../logger';
import { readLocalFile } from '../../../util/fs';
import { LockFileEntry } from './common';

export type YarnLock = Record<string, string>;

export async function getYarnLock(filePath: string): Promise<YarnLock> {
  const yarnLockRaw = await readLocalFile(filePath, 'utf8');
  try {
    const yarnLockParsed = parse(yarnLockRaw);
    // istanbul ignore if
    if (yarnLockParsed.type !== 'success') {
      logger.debug(
        { filePath, parseType: yarnLockParsed.type },
        'Error parsing yarn.lock - not success'
      );
      return {};
    }
    const lockFile: YarnLock = {};

    for (const [entry, val] of Object.entries(
      yarnLockParsed.object as LockFileEntry
    )) {
      logger.trace({ entry, version: val.version });
      lockFile[entry] = val.version;
    }
    return lockFile;
  } catch (err) {
    logger.debug({ filePath, err }, 'Warning: Exception parsing yarn.lock');
    return {};
  }
}
