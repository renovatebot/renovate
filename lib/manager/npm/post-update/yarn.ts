import { readFile } from 'fs-extra';
import { getInstalledPath } from 'get-installed-path';
import { join } from 'upath';
import { SYSTEM_INSUFFICIENT_DISK_SPACE } from '../../../constants/error-messages';
import { DatasourceError } from '../../../datasource';
import { logger } from '../../../logger';
import { exec } from '../../../util/exec';
import { BinarySource } from '../../../util/exec/common';
import { api as semver } from '../../../versioning/semver';
import { PostUpdateConfig, Upgrade } from '../../common';

export interface GenerateLockFileResult {
  error?: boolean;
  lockFile?: string;
  stderr?: string;
}

export async function generateLockFile(
  cwd: string,
  env?: NodeJS.ProcessEnv,
  config: PostUpdateConfig = {},
  upgrades: Upgrade[] = []
): Promise<GenerateLockFileResult> {
  const { binarySource } = config;
  logger.debug(`Spawning yarn install to create ${cwd}/yarn.lock`);
  let lockFile = null;
  let stdout = '';
  let stderr = '';
  let cmd: string;
  try {
    try {
      // See if renovate is installed locally
      const installedPath = join(
        await getInstalledPath('yarn', {
          local: true,
        }),
        'bin/yarn.js'
      );
      cmd = `node ${installedPath}`;
      const yarnIntegrity =
        config.upgrades &&
        config.upgrades.some((upgrade) => upgrade.yarnIntegrity);
      if (!yarnIntegrity) {
        logger.warn('Using yarn@1.9.4 for install is deprecated');
        try {
          const renovatePath = await getInstalledPath('renovate', {
            local: true,
          });
          logger.debug('Using nested bundled yarn@1.9.4 for install');
          cmd = 'node ' + join(renovatePath, 'bin/yarn-1.9.4.js');
        } catch (err) {
          logger.debug('Using bundled yarn@1.9.4 for install');
          cmd = cmd.replace(
            'node_modules/yarn/bin/yarn.js',
            'bin/yarn-1.9.4.js'
          );
        }
      }
    } catch (localerr) {
      logger.debug('No locally installed yarn found');
      // Look inside globally installed renovate
      try {
        const renovateLocation = await getInstalledPath('renovate');
        const installedPath = join(
          await getInstalledPath('yarn', {
            local: true,
            cwd: renovateLocation,
          }),
          'bin/yarn.js'
        );
        cmd = `node ${installedPath}`;
      } catch (nestederr) {
        logger.debug('Could not find globally nested yarn');
        // look for global yarn
        try {
          const installedPath = join(
            await getInstalledPath('yarn'),
            'bin/yarn.js'
          );
          cmd = `node ${installedPath}`;
        } catch (globalerr) {
          logger.warn('Could not find globally installed yarn');
          cmd = 'yarn';
        }
      }
    }
    if (binarySource === BinarySource.Global) {
      cmd = 'yarn';
    }

    const { stdout: yarnVersion } = await exec(`${cmd} --version`);

    logger.debug(`Using yarn: ${cmd} ${yarnVersion}`);

    const yarnMajorVersion = semver.getMajor(yarnVersion);

    let cmdExtras = '';
    const cmdEnv = { ...env };
    if (yarnMajorVersion < 2) {
      cmdExtras += ' --ignore-scripts';
    } else {
      cmdEnv.YARN_ENABLE_SCRIPTS = '0';
    }
    cmdExtras += ' --ignore-engines';
    cmdExtras += ' --ignore-platform';
    const installCmd = cmd + ' install' + cmdExtras;
    // TODO: Switch to native util.promisify once using only node 8
    await exec(installCmd, {
      cwd,
      env: cmdEnv,
    });
    const lockUpdates = upgrades
      .filter((upgrade) => upgrade.isLockfileUpdate)
      .map((upgrade) => upgrade.depName);
    if (lockUpdates.length) {
      logger.debug('Performing lockfileUpdate (yarn)');
      const updateCmd =
        cmd +
        ' upgrade' +
        lockUpdates.map((depName) => ` ${depName}`).join('') +
        cmdExtras;
      const updateRes = await exec(updateCmd, {
        cwd,
        env,
      });
      // istanbul ignore next
      stdout += updateRes.stdout || '';
      stderr += updateRes.stderr || '';
    }

    if (yarnMajorVersion < 2) {
      if (
        config.postUpdateOptions &&
        config.postUpdateOptions.includes('yarnDedupeFewer')
      ) {
        logger.debug('Performing yarn dedupe fewer');
        const dedupeCommand =
          'npx yarn-deduplicate@1.1.1 --strategy fewer && yarn';
        const dedupeRes = await exec(dedupeCommand, {
          cwd,
          env,
        });
        // istanbul ignore next
        stdout += dedupeRes.stdout || '';
        stderr += dedupeRes.stderr || '';
      }
      if (
        config.postUpdateOptions &&
        config.postUpdateOptions.includes('yarnDedupeHighest')
      ) {
        logger.debug('Performing yarn dedupe highest');
        const dedupeCommand =
          'npx yarn-deduplicate@1.1.1 --strategy highest && yarn';
        const dedupeRes = await exec(dedupeCommand, {
          cwd,
          env,
        });
        // istanbul ignore next
        stdout += dedupeRes.stdout || '';
        stderr += dedupeRes.stderr || '';
      }
    } else if (
      config.postUpdateOptions &&
      config.postUpdateOptions.some((option) => option.startsWith('yarnDedupe'))
    ) {
      logger.warn('yarn-deduplicate is not supported since yarn 2');
    }
    lockFile = await readFile(join(cwd, 'yarn.lock'), 'utf8');
  } catch (err) /* istanbul ignore next */ {
    logger.debug(
      {
        cmd,
        err,
        stdout,
        stderr,
        type: 'yarn',
      },
      'lock file error'
    );
    if (err.stderr) {
      if (err.stderr.includes('ENOSPC: no space left on device')) {
        throw new Error(SYSTEM_INSUFFICIENT_DISK_SPACE);
      }
      if (
        err.stderr.includes('The registry may be down.') ||
        err.stderr.includes('getaddrinfo ENOTFOUND registry.yarnpkg.com') ||
        err.stderr.includes('getaddrinfo ENOTFOUND registry.npmjs.org')
      ) {
        throw new DatasourceError(err);
      }
    }
    return { error: true, stderr: err.stderr };
  }
  return { lockFile };
}
