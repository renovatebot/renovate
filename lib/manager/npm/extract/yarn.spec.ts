import { readFileSync } from 'fs';
import { fs } from '../../../../test/util';
import { getYarnLock } from './yarn';

jest.mock('../../../util/fs');

describe('manager/npm/extract/yarn', () => {
  describe('.getYarnLock()', () => {
    it('returns empty if exception parsing', async () => {
      fs.readLocalFile.mockResolvedValueOnce('abcd');
      const res = await getYarnLock('package.json');
      expect(Object.keys(res)).toHaveLength(0);
    });
    it('extracts', async () => {
      const plocktest1Lock = readFileSync(
        'lib/manager/npm/__fixtures__/plocktest1/yarn.lock',
        'utf8'
      );
      fs.readLocalFile.mockResolvedValueOnce(plocktest1Lock);
      const res = await getYarnLock('package.json');
      expect(res).toMatchSnapshot();
      expect(Object.keys(res)).toHaveLength(7);
    });
  });
});
