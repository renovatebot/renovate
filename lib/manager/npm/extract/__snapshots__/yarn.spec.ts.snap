// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/npm/extract/yarn .getYarnLock() extracts 1`] = `
Object {
  "ansi-styles@^3.2.1": "3.2.1",
  "chalk@^2.4.1": "2.4.1",
  "color-convert@^1.9.0": "1.9.1",
  "color-name@^1.1.1": "1.1.3",
  "escape-string-regexp@^1.0.5": "1.0.5",
  "has-flag@^3.0.0": "3.0.0",
  "supports-color@^5.3.0": "5.4.0",
}
`;
